packages/web-pkg/src/services/loadingService.ts

```diff
@@ -0,0 +1,100 @@
+import * as uuid from 'uuid'
+import { eventBus } from 'web-pkg'
+import { debounce } from 'lodash-es'
+
+export enum LoadingEventTopics {
+  add = 'loading-service.add',
+  remove = 'loading-service.remove',
+  setProgress = 'loading-service.set-progress'
+}
+
+export interface LoadingTaskState {
+  total: number
+  current: number
+}
+
+export interface LoadingTask {
+  id: string
+  active: boolean
+  state?: LoadingTaskState
+}
+
+// time until a loading task is being set active
+const DEFAULT_DEBOUNCE_TIME = 200
+
+export class LoadingService {
+  private tasks: LoadingTask[] = []
+
+  public get isLoading(): boolean {
+    return this.tasks.some((e) => e.active)
+  }
+
+  /**
+   * Get the current progress from 0 to 100.
+   * Returns null if at least one task is indeterminate.
+   */
+  public get currentProgress(): number | null {
+    if (this.tasks.some((e) => !e.state && e.active)) {
+      return null
+    }
+
+    const tasks = this.tasks.filter((e) => !!e.state && e.active)
+    if (!tasks.length) {
+      return null
+    }
+
+    const progress = tasks.reduce((acc, task) => {
+      acc += task.state.current / task.state.total
+      return acc
+    }, 0)
+
+    return Math.round((progress / tasks.length) * 100)
+  }
+
+  public addTask(
+    callback: ({ setProgress }: { setProgress: (args: LoadingTaskState) => void }) => Promise<any>,
+    {
+      debounceTime = DEFAULT_DEBOUNCE_TIME,
+      indeterminate = true
+    }: { debounceTime?: number; indeterminate?: boolean } = {}
+  ): Promise<any> {
+    const task = {
+      id: uuid.v4(),
+      active: false,
+      ...(!indeterminate && { state: { total: 0, current: 0 } })
+    }
+    this.tasks.push(task)
+
+    const debounced = debounce(() => {
+      task.active = true
+      eventBus.publish(LoadingEventTopics.add)
+    }, debounceTime)
+    debounced()
+
+    const setProgress = ({ total, current }: LoadingTaskState) => {
+      if (!indeterminate) {
+        this.setProgress({ task, total, current })
+      }
+    }
+
+    return callback({ setProgress }).finally(() => {
+      this.removeTask(task.id)
+    })
+  }
+
+  private removeTask(id: string): void {
+    this.tasks = this.tasks.filter((e) => e.id !== id)
+    eventBus.publish(LoadingEventTopics.remove)
+  }
+
+  private setProgress({ task, total, current }): void {
+    if (!task.state) {
+      task.state = { total: 0, current: 0 }
+    }
+    task.state.total = total
+    task.state.current = current
+    eventBus.publish(LoadingEventTopics.setProgress)
+  }
+}
+
+export const loadingService = new LoadingService()
```

Comment:
I would prefer not to export a "default instance", but to make it available via `app.config.globalProperties.$loadingService` and to access that through a `useLoadingService` composable

Yes, I know we create these default instances for all or at least most of the other services, but I still think it's pretty ugly and we should stop 😅